quinn/src/send_stream.rs

```diff
@@ -130,64 +129,43 @@ impl SendStream {
         Poll::Ready(Ok(result))
     }
 
-    /// Shut down the send stream gracefully.
+    /// Notify the peer that no more data will ever be written to this stream
     ///
-    /// No new data may be written after calling this method. Completes when the peer has
-    /// acknowledged all sent data, retransmitting data as needed.
-    pub async fn finish(&mut self) -> Result<(), WriteError> {
-        Finish { stream: self }.await
-    }
-
-    /// Attempt to shut down the send stream gracefully.
+    /// It is an error to write to a [`SendStream`] after `finish`ing it. [`reset`](Self::reset) may
+    /// still be called after `finish` to abandon transmission of any stream data that might still
+    /// be buffered.
     ///
-    /// No new data may be written after calling this method. Completes when the peer has
-    /// acknowledged all sent data, retransmitting data as needed.
-    pub fn poll_finish(&mut self, cx: &mut Context) -> Poll<Result<(), WriteError>> {
-        let mut conn = self.conn.state.lock("poll_finish");
-        if self.is_0rtt {
-            conn.check_0rtt()
-                .map_err(|()| WriteError::ZeroRttRejected)?;
-        }
-        if self.finishing.is_none() {
-            conn.inner
-                .send_stream(self.stream)
-                .finish()
-                .map_err(|e| match e {
-                    FinishError::UnknownStream => WriteError::UnknownStream,
-                    FinishError::Stopped(error_code) => WriteError::Stopped(error_code),
-                })?;
-            let (send, recv) = oneshot::channel();
-            self.finishing = Some(recv);
-            conn.finishing.insert(self.stream, send);
-            conn.wake();
-        }
-        match Pin::new(self.finishing.as_mut().unwrap())
-            .poll(cx)
-            .map(|x| x.unwrap())
-        {
-            Poll::Ready(x) => {
-                self.finishing = None;
-                Poll::Ready(x.map_or(Ok(()), Err))
+    /// To wait for the peer to receive all buffered stream data, see [`stopped`](Self::stopped).
+    ///
+    /// May fail if [`finish`](Self::finish) or [`reset`](Self::reset) was previously called. This
+    /// error is harmless and serves only to indicate that the caller may have incorrect assumptions
+    /// about the stream's state.
+    pub fn finish(&mut self) -> Result<(), UnknownStream> {
+        let mut conn = self.conn.state.lock("finish");
+        match conn.inner.send_stream(self.stream).finish() {
+            Ok(()) => {
```

Comment:
Style: I'd maybe suggest making the `match` the tail expression of the function, and hoisting the `Ok(())` into the two arms, removing the explicit `return` in the third? Maybe also move the comment for the second arm onto instead of in the arm, which would help formatting conciseness.